crates/walrus-sui/src/client/retry_client/download_handler.rs

```diff
@@ -0,0 +1,599 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements a checkpoint download handler for the RPC client.
+use std::{
+    sync::{
+        Arc,
+        atomic::{AtomicU64, AtomicUsize, Ordering},
+    },
+    time::Instant,
+};
+
+use atomic_time::AtomicInstant;
+
+use crate::client::{retry_client::RetriableClientError, rpc_config::RpcFallbackConfig};
+
+/// The decision to make where to download a checkpoint.
+#[derive(Debug)]
+pub enum CheckpointDownloadDecision {
+    /// Attempt primary RPC first
+    AttemptPrimary,
+    /// Try fallback if primary fails
+    TryFallback(Option<RetriableClientError>),
+    /// Use primary error if fallback fails
+    UsePrimaryError(RetriableClientError),
```

Comment:
It looks like this case always leads to `unreachable!`. If that is the case, it might be beneficial to simply `panic!` upstream at the point where we would be returning this value anyway.